activerecord/lib/active_record/connection_adapters/abstract/transaction.rb

```diff
@@ -283,26 +283,24 @@ def rollback_transaction(transaction = nil)
 
       def within_new_transaction(options = {})
         @connection.lock.synchronize do
-          begin
-            transaction = begin_transaction options
-            yield
-          rescue Exception => error
-            if transaction
-              rollback_transaction
-              after_failure_actions(transaction, error)
-            end
-            raise
-          ensure
-            unless error
-              if Thread.current.status == "aborting"
-                rollback_transaction if transaction
-              else
-                begin
-                  commit_transaction if transaction
-                rescue Exception
-                  rollback_transaction(transaction) unless transaction.state.completed?
-                  raise
-                end
+          transaction = begin_transaction options
+          yield
+        rescue Exception => error
+          if transaction
+            rollback_transaction
+            after_failure_actions(transaction, error)
+          end
+          raise
+        ensure
+          unless error
+            if Thread.current.status == "aborting"
+              rollback_transaction if transaction
+            else
+              begin
```

Comment:
This `begin` is to rescue an exception raised by `commit_transaction` (`commit_transaction` is called when a transaction block is succeeded).

https://github.com/rails/rails/blob/49f9dff9b6ba1451d8c85927d5f75327bd2322d9/activerecord/test/cases/transactions_test.rb#L572-L585

This `begin` couldn't be removed since a begin could be ommited only when toplevel in methods and do-end blocks, but here is in if-else block.